nixos/modules/services/web-apps/immich.nix

```diff
@@ -0,0 +1,311 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.immich;
+  isPostgresUnixSocket = lib.hasPrefix "/" cfg.database.host;
+  isRedisUnixSocket = lib.hasPrefix "/" cfg.redis.host;
+
+  commonServiceConfig = {
+    Type = "simple";
+    Restart = "on-failure";
+    RestartSec = 3;
+
+    # Hardening
+    CapabilityBoundingSet = "";
+    NoNewPrivileges = true;
+    PrivateUsers = true;
+    PrivateTmp = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    RestrictAddressFamilies = [
+      "AF_INET"
+      "AF_INET6"
+      "AF_UNIX"
+    ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+  };
+  inherit (lib)
+    types
+    mkIf
+    mkOption
+    mkEnableOption
+    ;
+in
+{
+  options.services.immich = {
+    enable = mkEnableOption "Immich";
+    package = lib.mkPackageOption pkgs "immich" { };
```

Comment:
Do we need a `package` option?
According to the docs

> You shouldn’t necessarily make package options for all of your modules. You can always overwrite a specific package throughout nixpkgs by using [nixpkgs overlays](https://nixos.org/manual/nixpkgs/stable/#chap-overlays).

Reply:
Yes, overlays are really annoying to use. Every service should have a package option, I use them on the regular.

Reply:
Problem is that this one doesn't work like it should because it doesn't affect machine-learning.

Reply:
The systemd service uses `cfg.package.machine-learning` so you can easily provide a patched version for that as well

Reply:
Sure but you wouldn't expect
```
nix-repl> (immich.overrideAttrs (_: { version = "0"; })).machine-learning.version
"1.115.0"
```
I'll fix that in https://github.com/NixOS/nixpkgs/pull/344237 if you insist on keeping the option.

Reply:
Yes, i'd like to keep that option.